keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Multinomial.java

```diff
@@ -1,116 +1,92 @@
+        );
 
-    /**
-     * This method is necessary because I've constructed a flat array by concatenation samples of size k
-     * So for example, if the shape of n is [a, b]
-     * then I've now got a tensor of shape [a, b, k]
-     * which I need to convert to a tensor of shape [k, a, b]
-     * by doing a slice in the highest dimension and then concatenating again
-     *
-     * @param shape   the desired shape, not including the probabilities dimension
-     * @param samples the flat array of samples
-     * @return
-     */
-    private IntegerTensor constructSampleTensor(long[] shape, int[] samples) {
-        long[] outputShape = shape;
-        if (shape[0] == 1) {
-            outputShape = ArrayUtils.remove(outputShape, 0);
-        }
-        IntegerTensor abkTensor = IntegerTensor.create(samples, ArrayUtils.add(outputShape, numCategories));
-        int[] kabArray = new int[]{};
-        for (int category = 0; category < numCategories; category++) {
-            IntegerTensor abTensor = abkTensor.slice(outputShape.length, category);
-            kabArray = ArrayUtils.addAll(kabArray, abTensor.asFlatIntegerArray());
+        DoubleTensor broadcastedP = p.plus(DoubleTensor.zeros(broadcastResultShape));
+
+        double[] flatP = broadcastedP.asFlatDoubleArray();
+        int[] flatN = broadcastedN.asFlatIntegerArray();
+
+        int sampleCount = flatN.length;
+        int[] samples = new int[numCategories * sampleCount];
+
+        for (int i = 0; i < sampleCount; i++) {
+            int pIndex = i * numCategories;
+            drawNTimes(flatN[i], random, samples, pIndex, flatP, pIndex, numCategories);
         }
 
-        return IntegerTensor.create(kabArray, ArrayUtils.insert(0, outputShape, numCategories));
+        return IntegerTensor.create(samples, shape);
     }
 
-    private static int[] drawNTimes(int n, KeanuRandom random, Double... categoryProbabilities) {
-        int[] categoryDrawCounts = new int[categoryProbabilities.length];
+    private static void drawNTimes(int n, KeanuRandom random, int[] sample, int sampleIndex, double[] p, int pIndex, int pCount) {
         for (int i = 0; i < n; i++) {
-            int index = draw(random, categoryProbabilities);
-            categoryDrawCounts[index] += 1;
+            int index = draw(random, p, pIndex, pCount);
+            sample[sampleIndex + index] += 1;
         }
-        return categoryDrawCounts;
     }
 
-    private static int draw(KeanuRandom random, Double... categoryProbabilities) {
+    private static int draw(KeanuRandom random, double[] p, int pIndex, int pCount) {
         double value = random.nextDouble();
         int index = 0;
         Double pCumulative = 0.;
-        while (index < categoryProbabilities.length) {
-            Double currentP = categoryProbabilities[index++];
+        while (index < pCount) {
+            Double currentP = p[pIndex + index++];
```

Comment:
this is carnage can you do the incrementing separately - and why are we using boxed types?